datafusion/optimizer/src/expr_simplifier.rs

```diff
@@ -76,24 +96,181 @@ impl ExprSimplifiable for Expr {
     ///   }
     /// }
     ///
+    /// // Create the simplifier
+    /// let simplifier = ExprSimplifier::new(Info::default());
+    ///
     /// // b < 2
     /// let b_lt_2 = col("b").gt(lit(2));
     ///
     /// // (b < 2) OR (b < 2)
     /// let expr = b_lt_2.clone().or(b_lt_2.clone());
     ///
     /// // (b < 2) OR (b < 2) --> (b < 2)
-    /// let expr = expr.simplify(&Info::default()).unwrap();
+    /// let expr = simplifier.simplify(expr).unwrap();
     /// assert_eq!(expr, b_lt_2);
     /// ```
-    fn simplify<S: SimplifyInfo>(self, info: &S) -> Result<Self> {
-        let mut rewriter = Simplifier::new(info);
-        let mut const_evaluator = ConstEvaluator::try_new(info.execution_props())?;
+    pub fn simplify(&self, expr: Expr) -> Result<Expr> {
+        let mut rewriter = Simplifier::new(&self.info);
+        let mut const_evaluator = ConstEvaluator::try_new(self.info.execution_props())?;
 
         // TODO iterate until no changes are made during rewrite
         // (evaluating constants can enable new simplifications and
         // simplifications can enable new constant evaluation)
         // https://github.com/apache/arrow-datafusion/issues/1160
-        self.rewrite(&mut const_evaluator)?.rewrite(&mut rewriter)
+        expr.rewrite(&mut const_evaluator)?.rewrite(&mut rewriter)
+    }
+
+    /// Apply type coercion to an [`Expr`] so that it can be
+    /// evaluated as a [`PhysicalExpr`](datafusion_physical_expr::PhysicalExpr).
+    ///
+    /// See the [type coercion module](datafusion_expr::type_coercion)
+    /// documentation for more details on type coercion
+    ///
+    // Would be nice if this API could use the SimplifyInfo
```

Comment:
If this PR is approved, I will file a ticket to remove the `schema` argument in this function call.

Reply:
https://github.com/apache/arrow-datafusion/issues/3793